File: [0309][Medium][Best_Time_to_Buy_and_Sell_Stock_with_cooldown]/Best_Time_to_Buy_and_Sell_Stock_with_cooldown.py
# solution 1, one time traversal of the price list.
# leetcode time     cost : 44 ms
# leetcode memory   cost : 13.2 MB 
# Time  Complexity: O(N)
# Space Complexity: O(1)
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        n = len(prices)
        if not prices:
            return 0        
        dp_i_0 ,dp_i_1 =  0, -prices[0]
        dp_pre_2_0 = 0;       # represents dp[i-2][0]
        for i in range(0,n):
            dp_pre_1_0 = dp_i_0     # represents dp[i-1][0]
            dp_i_0 = max(dp_i_0, dp_i_1 + prices[i])
            dp_i_1 = max(dp_i_1, dp_pre_2_0 - prices[i])
            dp_pre_2_0 = dp_pre_1_0
        return dp_i_0; 
'''  
when k = +infinity  
dp[i][k][0] = max(dp[i-1][k][0], dp[i-1][k][1] + prices[i])
dp[i][k][1] = max(dp[i-1][k][1], dp[i-1][k-1][0] - prices[i])
            = max(dp[i-1][k][1], dp[i-1][k][0] - prices[i])
since k will not change, no need to record k,
dp[i][0] = max(dp[i-1][0], dp[i-1][1] + prices[i])
dp[i][1] = max(dp[i-1][1], dp[i-1][0] - prices[i])
'''
def main():  
    nums1 = [5,7,2,7,3,3,5,3,0] #expect is 9
    nums2 = [3,2,6,5,0,3] #expect is 7    
    obj = Solution()
    result = obj.maxProfit(nums2)
    print("return result is "+str(result));
    
if __name__ =='__main__':
    main() 

File: [0309][Medium][Best_Time_to_Buy_and_Sell_Stock_with_cooldown]/Best_Time_to_Buy_and_Sell_Stock_with_cooldown.java
// solution 1, brute force calculation.
// leetcode time     cost : 1 ms
// leetcode memory   cost : 37.1 MB 
// Time  Complexity: O(N)
// Space Complexity: O(1)
import java.math.*;
class Solution {
    public int maxProfit(int prices[]) {
        int n = prices.length;
        if (n == 0) 
            return 0;        
        int dp_i_0 = 0, dp_i_1 = -prices[0];
        int dp_pre_2_0 = 0;       // represents dp[i-2][0]
        // need to pay attention to index i's start(0) point and end point(n-1).
        for (int i = 0; i < prices.length; i++) {
            int dp_pre_1_0 = dp_i_0;     // represents dp[i-1][0]
            dp_i_0 = Math.max(dp_i_0, dp_i_1 + prices[i]);
            dp_i_1 = Math.max(dp_i_1, dp_pre_2_0 - prices[i]);
            dp_pre_2_0 = dp_pre_1_0;
        }
        return dp_i_0;
    }
}

public class Best_Time_to_Buy_and_Sell_Stock_with_cooldown {
    public static void main(String args[]) {
        int prices[] = {1,3,5,0,0,3,1,4}; // #expect is 8
        int prices_2[] = {3,2,6,5,0,3}; // #expect is 7
        Solution Solution_obj = new Solution();
        int result = Solution_obj.maxProfit(prices);
        System.out.println("In java code, 1st result is :" + result );
        int result2 = Solution_obj.maxProfit(prices_2);
        System.out.println("In java code, 2nd result is "+ result2);
    }
}